### ISSUE TITLE
Triton Inference Slower Than Torch On RTX 3090('sm_86' is not a recognized processor for this target (ignoring processor))

### ISSUE THREAD

**Issue by mallorbc** [mosaicml/llm-foundry#266]
## Environment
Package                  Version        Editable project location
------------------------ -------------- -------------------------
accelerate               0.19.0
aiohttp                  3.8.4
aiosignal                1.3.1
aniso8601                9.0.1
antlr4-python3-runtime   4.9.3
anyio                    3.7.0
apache-libcloud          3.7.0
appdirs                  1.4.4
argcomplete              3.0.8
arrow                    1.2.3
async-timeout            4.0.2
asyncio                  3.4.3
attr                     0.3.2
attrs                    23.1.0
backoff                  2.2.1
bcrypt                   4.0.1
bitsandbytes             0.39.0
blinker                  1.6.2
boto3                    1.26.145
botocore                 1.29.145
Brotli                   1.0.9
certifi                  2023.5.7
cffi                     1.15.1
cformers                 0.0.4
charset-normalizer       3.1.0
circuitbreaker           1.4.0
click                    8.1.3
cmake                    3.26.3
coloredlogs              15.0.1
composer                 0.14.1
contourpy                1.0.7
coolname                 2.2.0
cryptography             39.0.2
cycler                   0.11.0
datasets                 2.10.1
decorator                5.1.1
deepspeed                0.9.2+e0e80850
deepspeed-mii            0.0.5+e440f83
dill                     0.3.6
docker                   6.1.3
docker-pycreds           0.4.0
einops                   0.5.0
evaluate                 0.4.0
exceptiongroup           1.1.1
fastapi                  0.95.2
filelock                 3.12.0
flash-attn               1.0.3.post0
Flask                    2.3.2
Flask-RESTful            0.3.10
flatbuffers              23.5.26
fonttools                4.39.4
frozenlist               1.3.3
fsspec                   2023.5.0
gitdb                    4.0.10
GitPython                3.1.31
gql                      3.4.1
graphql-core             3.2.3
grpcio                   1.54.2
grpcio-tools             1.54.2
h11                      0.14.0
hjson                    3.1.0
huggingface-hub          0.14.1
humanfriendly            10.0
idna                     3.4
importlib-metadata       6.6.0
itsdangerous             2.1.2
Jinja2                   3.1.2
jmespath                 1.0.1
kiwisolver               1.4.4
lit                      16.0.5
llm-foundry              0.1.0          /tmp/llm-foundry
markdown-it-py           2.2.0
MarkupSafe               2.1.2
matplotlib               3.7.1
mdurl                    0.1.2
mosaicml-cli             0.4.5
mosaicml-streaming       0.4.1
mpmath                   1.3.0
multidict                6.0.4
multiprocess             0.70.14
mypy-extensions          1.0.0
nest-asyncio             1.5.6
networkx                 3.1
ninja                    1.11.1
numpy                    1.24.3
nvidia-cublas-cu11       11.10.3.66
nvidia-cuda-cupti-cu11   11.7.101
nvidia-cuda-nvrtc-cu11   11.7.99
nvidia-cuda-runtime-cu11 11.7.99
nvidia-cudnn-cu11        8.5.0.96
nvidia-cufft-cu11        10.9.0.58
nvidia-curand-cu11       10.2.10.91
nvidia-cusolver-cu11     11.4.0.1
nvidia-cusparse-cu11     11.7.4.91
nvidia-nccl-cu11         2.14.3
nvidia-nvtx-cu11         11.7.91
nvidia-pyindex           1.0.9
oci                      2.104.0
omegaconf                2.3.0
onnx                     1.13.1
onnxruntime              1.14.1
onnxruntime-gpu          1.15.0
optimum                  1.8.7.dev0
packaging                22.0
pandas                   2.0.2
paramiko                 3.2.0
pathtools                0.1.2
Pillow                   9.5.0
pip                      22.0.2
polygraphy               0.47.1
prompt-toolkit           3.0.38
protobuf                 3.20.3
psutil                   5.9.5
py-cpuinfo               9.0.0
pyarrow                  12.0.0
pycparser                2.21
pydantic                 1.10.8
Pygments                 2.15.1
PyNaCl                   1.5.0
pyOpenSSL                23.2.0
pyparsing                3.0.9
pyre-extensions          0.0.29
python-dateutil          2.8.2
python-snappy            0.6.1
pytorch-ranger           0.1.1
pytz                     2023.3
PyYAML                   6.0
questionary              1.10.0
redis                    4.5.5
regex                    2023.5.5
requests                 2.31.0
responses                0.18.0
rich                     13.4.1
ruamel.yaml              0.17.31
ruamel.yaml.clib         0.2.7
s3transfer               0.6.1
safetensors              0.3.1
sentencepiece            0.1.97
sentry-sdk               1.24.0
setproctitle             1.3.2
setuptools               59.6.0
six                      1.16.0
slack-sdk                3.21.3
smmap                    5.0.0
sniffio                  1.3.0
starlette                0.27.0
sympy                    1.12
tabulate                 0.9.0
timm                     0.9.2
tokenizers               0.13.3
torch                    2.0.1
torch-optimizer          0.3.0
torchaudio               2.0.2
torchdata                0.6.1
torchmetrics             0.11.3
torchtext                0.15.2
torchvision              0.15.2
tqdm                     4.65.0
transformers             4.28.1
triton                   2.0.0
triton-pre-mlir          2.0.0
typing_extensions        4.6.2
typing-inspect           0.9.0
tzdata                   2023.3
Unidecode                1.3.6
urllib3                  1.26.16
uvicorn                  0.22.0
validators               0.20.0
wandb                    0.15.3
wcwidth                  0.2.6
websocket-client         1.5.2
websockets               10.4
Werkzeug                 2.3.4
wget                     3.2
wheel                    0.37.1
xentropy-cuda-lib        0.1
xformers                 0.0.20
xxhash                   3.2.0
yarl                     1.9.2
zipp                     3.15.0
zstd                     1.5.5.1

I am also using RTX 3090 inside of a Cuda 11.7 docker container
## To reproduce

1. Install llm-foundry from source using README and install for GPUs
2. Try using Triton attention as well as stock torch attention
3. Generate a large number of tokens, say 2048 tokens, and notice that time to generate is larger for Triton.

I am seeing 45 seconds for torch and 48-52 for triton.  I am turning use_cache to True, as it is False by default, not sure if that makes a difference.  If left to False both are MUCH slower as expected

## Expected behavior

Triton is supposed to be faster.  It does not seem that it is.

## Additional context

The first time inferencing using Triton it says something like this: 'sm_86' is not a recognized processor for this target (ignoring processor)

This is a message I have seen in the past I believe when my 3090s where new and the hardware was not yet supported.  Considering that RTX 3090s are coming up on 3 years old, is this still the case?

```python
import transformers
import time
import torch
import argparse

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-t","--use_triton",action="store_true",help="Use triton")
    args = parser.parse_args()
    config = transformers.AutoConfig.from_pretrained(
    'mosaicml/mpt-7b',
    trust_remote_code=True
    )
    if args.use_triton:
        config.attn_config['attn_impl'] = 'triton'
    config.use_cache = True

    model = transformers.AutoModelForCausalLM.from_pretrained(
    'mosaicml/mpt-7b',
    config=config,
    torch_dtype=torch.bfloat16,
    trust_remote_code=True
    )
    model.to(device='cuda:0')
    tokenizer = transformers.AutoTokenizer.from_pretrained("mosaicml/mpt-7b")
    print("Model loaded")
    model.eval()
    start = time.time()
    input_ids = tokenizer.encode("Hello", return_tensors="pt").cuda()
    output = model.generate(input_ids, max_length=2048, do_sample=False, num_return_sequences=1)
    print(f"Time taken: {time.time()-start}")
```

**Comment by vchiley** [comment]
`'sm_86' is not a recognized processor for this target (ignoring processor)`
is an inconsequential warning
This [note is being added to our FAQ](https://github.com/mosaicml/llm-foundry/blob/tutorial/TUTORIAL.md#known-issue-with-sm89-gpus) (where the FAQ should be corrected to say sm86+ GPUs)

**Comment by vchiley** [comment]
The comments made [here](https://github.com/mosaicml/llm-foundry/blob/tutorial/TUTORIAL.md#what-are-the-different-attention-options-torch--flash--triton--for-mpt-and-which-one-should-i-use) about throughput improvements are training specific. I'm not 100% sure about the inference implications of the different attention implementations.

**Comment by mallorbc** [comment]
Ah, I had the wrong impression.  There is no impact on inference and so what I am seeing is normal.  Some other comments suggest that there may be a difference for large inputs.  I will share those results and then close this issue.

Thanks!

**Comment by mallorbc** [comment]
Using an input of 1500 tokens and generating the remaining 548, I got a generation time of 14.4 seconds for torch implementation and a time of 16 seconds when using Triton implementation.

The only other thought that I have is that while is seems Triton is slower than Torch for inference, perhaps it uses less memory for inference as well(because it does while training).  Perhaps this would be useful then for generating really long sequences where there are not enough resources otherwise.

Thanks!
